src/common/io/http/url.c

```diff
@@ -0,0 +1,213 @@
+/***********************************************************************************************************************************
+HTTP URL
+***********************************************************************************************************************************/
+#include "build.auto.h"
+
+#include "common/debug.h"
+#include "common/io/http/url.h"
+#include "common/memContext.h"
+#include "common/type/object.h"
+#include "common/type/stringList.h"
+#include "common/regExp.h"
+
+/***********************************************************************************************************************************
+Regular expression for URLs. This is not intended to be completely comprehensive, e.g. it is still possible to enter bad hostnames.
+The goal is to make sure the syntax is correct enough for the rest of the parsing to succeed.
+***********************************************************************************************************************************/
+STRING_STATIC(
+    HTTP_URL_REGEXP_STR,
+    "^(http[s]{0,1}:\\/\\/){0,1}"                                   // Optional protocol (http or https)
+    "([^\\[\\:\\/?]+|\\[[a-fA-F0-9:]+\\])"                          // host/ipv4/ipv6
+    "(:[1-9][0-9]{0,4}){0,1}"                                       // Optional port
+    "(\\/[^?\\/]*)*$");                                             // Optional path
+
+/***********************************************************************************************************************************
+Object type
+***********************************************************************************************************************************/
+struct HttpUrl
+{
+    HttpUrlPub pub;
+    MemContext *memContext;                                         // Mem context
```

Comment:
Is the MemContext below the pub for alignment? I just found it strange that MemContext was not the first thing in the struct as it is everywhere else, and the way it's constructed in lines 75-82, so I was curious as to why.

Reply:
It needs to be first so we know where to find it, i.e. when casting from HttpUrl to HttpUrlPub the pub elements need to be at the beginning of the private struct.

It's the same for String, the MemContext comes after the STRING_COMMON define.